### PR TITLE
Implements patch discriminator

### PR DESCRIPTION
**Pull request by callumnewlands**

### CODE REVIEW

File: patch_discriminator.py [hunk 1]
```diff
@@ -10,25 +10,24 @@ class PatchDiscriminator(nn.Module):
...
         num_down_samples = int(math.ceil(math.log(patch_dim, 2)))
         scale_channel_amount = int(self.scale_up_to / (2 ** num_down_samples) * channel_scale_factor)
 
-        current_dim = patch_dim
-
         downscalers = []
         in_channels = scale_channel_amount
-        while current_dim > 4:
+        for sample in range(num_down_samples, 2, -1):
             out_channels = min(in_channels * 2, self.mc)
```
Comment (jacob-scott3141) [inline patch_discriminator.py:30]: Sorry for perhaps a silly question, but why is this multiplied by 2 here?

File: patch_discriminator.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            # Average encodings over reference_patches (possibly in batches)
+            # Not 100% sure this is correct but when it's integrated with real data that should let us know
+            b, n, *_ = reference_patches.size()
+            if reference_patches.ndim == 5:  # Batch of b sets of n patches (from b images)
+                flattened_patches = reference_patches.flatten(0, 1)
+            else:
+                flattened_patches = reference_patches
+            encoded_references = self.patchEncoder(flattened_patches)
+            _, c, h, w = encoded_references.size()
+            encoded_references = encoded_references.view(b, n, c, h, w)
```
Comment (jacob-scott3141) [inline patch_discriminator.py:69]: I might need to go through the matrix shapes on paper to understand where b, n, c, h, and w come from

**Review by jacob-scott3141 — APPROVED** [review]
LGTM, I don't understand a few of the lines (where I've commented), but the rest seems to make sense.

File: patch_discriminator.py (hunk 1 above)
Comment (callumnewlands) [reply patch_discriminator.py:30]: It doubles the number of channels each time to reduce the layer dimensions by half
So a 256x256x2 input => 128x128x4 => 64x64x8 => 32x32x16 => 16x16x32 => ...

File: patch_discriminator.py (hunk 2 above)
Comment (callumnewlands) [reply patch_discriminator.py:69]: I copied it from the way they did it to match their training data shapes, but I think the first dimension of a network is usually the batch size b, and the last 3 are channels, height and width (c, h, w) which is the standard tensor way of representing images, and then I think n here they're using as a the number of patches to do at once (n patches taken from  each image)?

And I think this allows us to use it for the average encoding over n patches, as well as performing a single encoding on a given patch -- but I'm not 100% sure so we'll probably have to wait to integrate it with the actual data and see if the shapes match up

**Pull request merged**